sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/ProxyAuthenticationHandler.java

```diff
@@ -0,0 +1,287 @@
+        List<Map<String, String>> digestChallenges = new ArrayList<>();
+
+        for (String proxyAuthenticationHeader : headers.getAll(PROXY_AUTHENTICATE)) {
+            String[] typeValuePair = proxyAuthenticationHeader.split(" ", 2);
+
+            String challengeType = typeValuePair[0].trim();
+            if (challengeType.equalsIgnoreCase(AUTH_BASIC)) {
+                /*
+                 * Proxy-Authenticate is requesting Basic authorization, this only needs a flag as Basic authentication
+                 * is always the same.
+                 */
+                hasBasicChallenge = true;
+            } else if (challengeType.equalsIgnoreCase(AUTH_DIGEST)) {
+                /*
+                 * Proxy-Authenticate is requesting Digest authorization, this needs to be parsed for the challenge
+                 * information as Digest authentication always changes.
+                 */
+                Map<String, String> digestChallenge = new HashMap<>();
+                for (String challengePiece : typeValuePair[1].split(",")) {
+                    String[] kvp = challengePiece.split("=", 2);
+
+                    // Skip challenge information that has no value.
+                    if (kvp.length != 2) {
+                        continue;
+                    }
+
+                    digestChallenge.put(kvp[0].trim(), kvp[1].trim().replace("\"", ""));
+                }
+
+                digestChallenges.add(digestChallenge);
+            }
+        }
+
+        return new ChallengeHolder(hasBasicChallenge, digestChallenges);
+    }
+
+    /*
+     * Validate the Proxy-Authorization header used in authentication against the server's Proxy-Authentication-Info
+     * header. This header is an optional return by the server so this validation is guaranteed to happen.
+     */
+    private void validateProxyAuthenticationInfo(String infoHeader, String authorizationHeader) {
+        // Server didn't return a 'Proxy-Authentication-Info' header, nothing to consume.
+        if (CoreUtils.isNullOrEmpty(infoHeader)) {
+            return;
+        }
+
+        Map<String, String> authenticationInfoPieces = AuthorizationChallengeHandler
+            .parseAuthenticationOrAuthorizationHeader(infoHeader);
+        Map<String, String> authorizationPieces = AuthorizationChallengeHandler
+            .parseAuthenticationOrAuthorizationHeader(authorizationHeader);
+
+        /*
+         * If the authentication info response contains a cnonce or nc value it MUST match the value sent in the
+         * authorization header. This is the server performing validation to the client that it received the
+         * information.
+         */
+        assert !authenticationInfoPieces.containsKey(CNONCE)
+            || authenticationInfoPieces.get(CNONCE).equals(authorizationPieces.get(CNONCE));
+        assert !authenticationInfoPieces.containsKey(NC)
+            || authenticationInfoPieces.get(NC).equals(authorizationPieces.get(NC));
```

Comment:
Prefer to throw exceptions instead of using `assert`. If these assertions fail, they throw `AssertionError` which is of type [Error](https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html) and is considered a serious non-recoverable problem with the application and is not expected to be handled upstream.

```
An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it.
A method is not required to declare in its throws clause any subclasses of Error that might be thrown during the execution of the method but not caught, since these errors are abnormal conditions that should never occur. That is, Error and its subclasses are regarded as unchecked exceptions for the purposes of compile-time checking of exceptions.
```

Reply:
Updated to throw an exception instead.